File: examples/cpp/sample_discovery.cpp
/**
 * @file sample_discovery.cpp
 * @brief Sample: Discover AFI920 sensors on the network (multi-LAN)
 *
 * Enumerates all local network interfaces and sends a UDP broadcast
 * Discovery request from each one, collecting responses across all LANs.
 *
 * Usage: sample_discovery [timeout_ms] [bind_ip]
 *
 *   If bind_ip is given, discovery is limited to that single interface.
 *   Otherwise, all non-loopback interfaces are scanned automatically.
 */
#include <afi/afi_sdk.h>
#include <cstdio>
#include <cstdlib>
#include <set>
#include <string>
#include <vector>

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#include <iphlpapi.h>
#pragma comment(lib, "iphlpapi.lib")
#pragma comment(lib, "ws2_32.lib")
#else
#include <arpa/inet.h>
#include <ifaddrs.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <unistd.h>
#endif

struct LocalInterface {
    std::string ip;
    std::string name;
};

/// Enumerate local non-loopback IPv4 interfaces
static std::vector<LocalInterface> enumerate_local_ipv4() {
    std::vector<LocalInterface> results;

#ifdef _WIN32
    ULONG size = 15000;
    PIP_ADAPTER_ADDRESSES addrs = (PIP_ADAPTER_ADDRESSES)malloc(size);
    if (GetAdaptersAddresses(AF_INET, 0, nullptr, addrs, &size) == ERROR_SUCCESS) {
        for (auto a = addrs; a; a = a->Next) {
            if (a->OperStatus != IfOperStatusUp) continue;
            for (auto u = a->FirstUnicastAddress; u; u = u->Next) {
                if (u->Address.lpSockaddr->sa_family == AF_INET) {
                    auto* sa = (sockaddr_in*)u->Address.lpSockaddr;
                    char buf[INET_ADDRSTRLEN];
                    inet_ntop(AF_INET, &sa->sin_addr, buf, sizeof(buf));
                    std::string ip(buf);
                    if (ip.substr(0, 4) != "127.") {
                        char name_buf[256];
                        WideCharToMultiByte(CP_UTF8, 0, a->FriendlyName, -1,
                                            name_buf, sizeof(name_buf), nullptr, nullptr);
                        results.push_back({ip, name_buf});
                    }
                }
            }
        }
    }
    free(addrs);
#else
    struct ifaddrs* ifaddr = nullptr;
    if (getifaddrs(&ifaddr) == 0) {
        for (auto* ifa = ifaddr; ifa; ifa = ifa->ifa_next) {
            if (!ifa->ifa_addr) continue;
            if (ifa->ifa_addr->sa_family != AF_INET) continue;
            if (ifa->ifa_flags & IFF_LOOPBACK) continue;

            auto* sa = reinterpret_cast<struct sockaddr_in*>(ifa->ifa_addr);
            char buf[INET_ADDRSTRLEN];
            inet_ntop(AF_INET, &sa->sin_addr, buf, sizeof(buf));
            results.push_back({buf, ifa->ifa_name ? ifa->ifa_name : ""});
        }
        freeifaddrs(ifaddr);
    }
#endif

    return results;
}

static void print_sensor(size_t idx, const afi::AfiSensorInfo& s) {
    printf("[Sensor %zu]\n", idx);
    printf("  IP:             %s:%d\n", s.ip.c_str(), s.port);
    printf("  Serial:         %s\n", s.serial_number.c_str());
    printf("  Model:          %s\n", s.model.c_str());
    printf("  SW Version:     %s (bl=%s drv=%s fw=%s)\n", s.software_version.c_str(),
           s.bootloader_hash.c_str(), s.driver_hash.c_str(), s.fw_hash.c_str());
    printf("  HW Version:     %s\n", s.hw_version.c_str());
    printf("  MAC:            %s\n", s.mac_address.c_str());
    printf("  Subnet:         %s\n", s.subnet_mask.c_str());
    printf("  Gateway:        %s\n", s.gateway.c_str());
    printf("  Range Mode:     %s (%d)\n", s.range_mode_name.c_str(),
           static_cast<int>(s.range_mode));
    printf("  Position Code:  %d\n", s.position_code);
    printf("  Mounting:       (%.3f, %.3f, %.3f) m\n",
           s.offset_x, s.offset_y, s.offset_z);
    printf("  Orientation:    yaw=%.1f pitch=%.1f roll=%.1f deg\n",
           s.yaw_deg, s.pitch_deg, s.roll_deg);
    printf("  Ports:          det=%d health=%d perf=%d\n",
           s.detection_port, s.health_port, s.performance_port);
    printf("  Host IP:        %s\n", s.host_ip.c_str());
    printf("  Det IP:         %s\n", s.detection_ip.c_str());
    printf("\n");
}

int main(int argc, char* argv[]) {
    int timeout_ms = 3000;
    std::string bind_ip;
    if (argc > 1) timeout_ms = std::atoi(argv[1]);
    if (argc > 2) bind_ip = argv[2];

    // Determine interfaces to scan
    std::vector<LocalInterface> interfaces;
    if (!bind_ip.empty()) {
        interfaces.push_back({bind_ip, "user-specified"});
    } else {
        interfaces = enumerate_local_ipv4();
        if (interfaces.empty()) {
            interfaces.push_back({"", "default"});
        }
    }

    printf("Discovering AFI920 sensors (timeout: %d ms, %zu interface(s))...\n\n",
           timeout_ms, interfaces.size());

    // Discover on each interface, dedup by ip+serial
    std::vector<afi::AfiSensorInfo> all_sensors;
    std::set<std::string> seen_keys;

    for (const auto& iface : interfaces) {
        std::string label = iface.name.empty()
                            ? (iface.ip.empty() ? "any" : iface.ip)
                            : iface.name + " (" + iface.ip + ")";
        printf("  Scanning on %s ...\n", label.c_str());

        auto found = afi::AfiSdk::Discover(timeout_ms, 30520, iface.ip);

        for (auto& s : found) {
            std::string key = s.ip + ":" + s.serial_number;
            if (seen_keys.find(key) == seen_keys.end()) {
                seen_keys.insert(key);
                all_sensors.push_back(std::move(s));
            }
        }

        printf("    -> %zu response(s)\n", found.size());
    }

    printf("\nTotal unique sensors found: %zu\n\n", all_sensors.size());

    if (all_sensors.empty()) {
        printf("No sensors found.\n");
        return 0;
    }

    for (size_t i = 0; i < all_sensors.size(); i++) {
        print_sensor(i + 1, all_sensors[i]);
    }

    return 0;
}

File: python/examples/sample_discovery.py
"""
Sample: Discover AFI920 sensors on the network (multi-LAN)

Enumerates all local network interfaces and sends a UDP broadcast
Discovery request from each one, collecting responses across all LANs.

Usage:
    python sample_discovery.py [timeout_sec] [bind_ip]

    If bind_ip is given, discovery is limited to that single interface.
    Otherwise, all non-loopback interfaces are scanned automatically.

Copyright 2026 bitsensing Inc. All rights reserved.
"""

import socket
import struct
import sys
from afi_sdk import AfiSdk, RangeMode


def enumerate_local_ipv4() -> list[dict[str, str]]:
    """Return local non-loopback IPv4 interfaces as [{"ip": ..., "name": ...}]."""
    seen: set[str] = set()
    results: list[dict[str, str]] = []

    # Linux: fcntl + SIOCGIFADDR (catches USB/external adapters)
    if sys.platform != "win32":
        try:
            import fcntl

            for _idx, name in socket.if_nameindex():
                try:
                    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                    ip_bytes = fcntl.ioctl(
                        s.fileno(), 0x8915,  # SIOCGIFADDR
                        struct.pack("256s", name.encode("utf-8")[:15]),
                    )[20:24]
                    s.close()
                    ip_str = socket.inet_ntoa(ip_bytes)
                    if not ip_str.startswith("127.") and ip_str not in seen:
                        seen.add(ip_str)
                        results.append({"ip": ip_str, "name": name})
                except OSError:
                    pass
        except (ImportError, OSError):
            pass

    # Fallback: getaddrinfo
    try:
        for info in socket.getaddrinfo(socket.gethostname(), None, socket.AF_INET):
            ip_str = str(info[4][0])
            if not ip_str.startswith("127.") and ip_str not in seen:
                seen.add(ip_str)
                results.append({"ip": ip_str, "name": ""})
    except OSError:
        pass

    return results


def print_sensor(idx: int, s: dict):
    """Print a single sensor's discovery info."""
    mode = s.get("range_mode", -1)
    print(f"[Sensor {idx}]")
    print(f"  IP:             {s.get('ip_address', s.get('_source_ip', '?'))}:"
          f"{s.get('api_port', s.get('_source_port', '?'))}")
    print(f"  Serial:         {s.get('serial_number', '?')}")
    print(f"  Model:          {s.get('model_name', '?')}")
    print(f"  SW Version:     {s.get('software_version', '?')} "
          f"(bl={s.get('bootloader_hash', '?')} drv={s.get('driver_hash', '?')} "
          f"fw={s.get('fw_hash', '?')})")
    print(f"  HW Version:     {s.get('hw_version', '?')}")
    print(f"  MAC:            {s.get('mac_address', '?')}")
    print(f"  Range Mode:     {s.get('range_mode_name', RangeMode.name(mode))} ({mode})")
    print(f"  Position Code:  {s.get('position_code', '?')}")
    print(f"  Mounting:       ({s.get('offset_x_m', 0):.3f}, "
          f"{s.get('offset_y_m', 0):.3f}, {s.get('offset_z_m', 0):.3f}) m")
    print(f"  Ports:          det={s.get('detection_port', '?')} "
          f"health={s.get('health_port', '?')} perf={s.get('performance_port', '?')}")
    scan = s.get("scan_status")
    if isinstance(scan, dict):
        print(f"  Scan:           index={scan.get('scan_index', '?')} "
              f"uptime={scan.get('uptime_seconds', '?')}s")
    if "ptp_state" in s:
        print(f"  PTP:            {s.get('ptp_state', '?')} "
              f"offset={s.get('ptp_offset_ns', '?')} ns")
    print()


def main():
    timeout = float(sys.argv[1]) if len(sys.argv) > 1 else 3.0
    bind_ip = sys.argv[2] if len(sys.argv) > 2 else ""

    # Determine interfaces to scan
    if bind_ip:
        interfaces = [{"ip": bind_ip, "name": "user-specified"}]
    else:
        interfaces = enumerate_local_ipv4()
        if not interfaces:
            interfaces = [{"ip": "", "name": "default"}]

    print(f"Discovering AFI920 sensors (timeout: {timeout}s, "
          f"{len(interfaces)} interface(s))...\n")

    # Discover on each interface, dedup by (ip, serial)
    all_sensors: list[dict] = []
    seen_keys: set[tuple[str, str]] = set()

    for iface in interfaces:
        ip = iface["ip"]
        name = iface.get("name", "")
        label = f"{name} ({ip})" if name else ip or "any"
        print(f"  Scanning on {label} ...")

        found = AfiSdk.discover(timeout_s=timeout, bind_ip=ip)

        for s in found:
            key = (
                s.get("ip_address", s.get("_source_ip", "")),
                s.get("serial_number", ""),
            )
            if key not in seen_keys:
                seen_keys.add(key)
                all_sensors.append(s)

        print(f"    -> {len(found)} response(s)")

    print(f"\nTotal unique sensors found: {len(all_sensors)}\n")

    if not all_sensors:
        print("No sensors found.")
        return

    for i, s in enumerate(all_sensors):
        print_sensor(i + 1, s)


if __name__ == "__main__":
    main()
